github/resource_github_repository_ruleset.go

```diff
@@ -0,0 +1,550 @@
+										Required:    true,
+										Description: "The pattern to match with.",
+									},
+								},
+							},
+						},
+						"tag_name_pattern": {
+							Type:        schema.TypeList,
+							MaxItems:    1,
+							Optional:    true,
+							Description: "Parameters to be used for the tag_name_pattern rule.",
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"enabled": {
+										Type:     schema.TypeBool,
+										Optional: true,
+										Default:  false,
+									},
+									"name": {
+										Type:        schema.TypeString,
+										Optional:    true,
+										Description: "How this rule will appear to users.",
+									},
+									"negate": {
+										Type:        schema.TypeBool,
+										Optional:    true,
+										Description: "If true, the rule will fail if the pattern matches.",
+									},
+									"operator": {
+										Type:        schema.TypeString,
+										Required:    true,
+										Description: "The operator to use for matching. Can be one of: `starts_with`, `ends_with`, `contains`, `regex`.",
+									},
+									"pattern": {
+										Type:        schema.TypeString,
+										Required:    true,
+										Description: "The pattern to match with.",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"etag": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+		// CustomizeDiff: customDiffFunction,
+	}
+}
+
+func resourceGithubRepositoryRulesetCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Owner).v3client
+
+	rulesetReq := resourceGithubRulesetObject(d, false)
+
+	owner := meta.(*Owner).name
+	repoName := d.Get("repository").(string)
+	ctx := context.Background()
+
+	var ruleset *github.Ruleset
+	var err error
+
+	ruleset, _, err = client.Repositories.CreateRuleset(ctx, owner, repoName, rulesetReq)
+	if err != nil {
+		return err
+	}
+	d.SetId(strconv.FormatInt(*ruleset.ID, 10))
+
+	return resourceGithubRepositoryRulesetRead(d, meta)
+}
+
+func resourceGithubRepositoryRulesetRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Owner).v3client
+
+	owner := meta.(*Owner).name
+	repoName := d.Get("repository").(string)
+	rulesetID, err := strconv.ParseInt(d.Id(), 10, 64)
+	if err != nil {
+		return unconvertibleIdErr(d.Id(), err)
+	}
+
+	ctx := context.WithValue(context.Background(), ctxId, rulesetID)
+	if !d.IsNewResource() {
+		ctx = context.WithValue(ctx, ctxEtag, d.Get("etag").(string))
+	}
+
+	var ruleset *github.Ruleset
+	var resp *github.Response
+
+	ruleset, resp, err = client.Repositories.GetRuleset(ctx, owner, repoName, rulesetID, false)
+	if err != nil {
+		if ghErr, ok := err.(*github.ErrorResponse); ok {
+			if ghErr.Response.StatusCode == http.StatusNotModified {
+				return nil
+			}
+			if ghErr.Response.StatusCode == http.StatusNotFound {
+				log.Printf("[INFO] Removing ruleset %s/%s: %d from state because it no longer exists in GitHub",
```

Comment:
I think this is a big nono in Terraform Providers. Automatically removing stuff from the state feels wrong. I think it should just crash, but I'm not sure what the best practices are here.

Reply:
This is something I took out of `resource_repository`, it’s essentially used for state refresh, if we can’t find it then it’s not there anymore and we update the state accordingly, at least that’s what I gathered from the current implementation of the provider

Reply:
Fair enough, I might be wrong on that point